### ISSUE TITLE
Parallel scopes

### ISSUE THREAD

**Issue by seesharper** [seesharper/LightInject#112]
Add support for parallel scopes that allows a scope to be started without automatically attaching to the ambient scope.

```
container.BeginScope(ScopeOptions.RequiresNew);
```

**Comment by prasannavl** [comment]
Hi, 

Wouldn't it be a better design approach be to use a scoping like Grace (https://github.com/ipjohnson/Grace/) does?

Where things are actually resolved from the scope instead of the container. (Resolving from the container resolves from the root scope). This way, you have explicit control over the created scopes, and easy to manage parallel scopes, and even scopes with custom registrations.

**Comment by seesharper** [comment]
I'm afraid that it would require a major rewrite with regards to the current implementation. That being said, the thought has crossed my mind that LightInject might need something like that in order to support parallel scopes. This is something that I definitively will look into, although the current scoping mechanism works quite well and very few people has requested "true" child container support.

**Comment by geirsagberg** [comment]
Ran into an issue where I would need parallel scopes today, when writing tests using multiple instances that each need their own scope, and that would complete at random times.

In other words, this feature would be nice to have :)

**Comment by seesharper** [comment]
LightInject supports parallell scopes in the sense that scopes are controlled by the current IScopeManagerProvider. This provider is by default set to the PerThreadScopeManagerProvider which means that there can be an hierarchy of scopes per thread.  In the world of async and await, there is the PerLogicalCallContextScopeManagerProvider that manages the scopes per "logical thread of execution" 
So given this information, we could argue that parallell scopes are already supported. Sort of anyways :)
The challenge with creating multiple parallell scopes within the same ScopeManager (provider by the IScopeManagerProvider) is that LightInject asks the ScopeManager for the current scope to be passed down to the ILifetime implementation that again controls the lifecycle of the instance created. If there were to be multiple possible current scopes that would certainly be a problem :) A better solution might be to retrieve the instance directly off the scope as discussed in #291.
You also might want to look at https://github.com/seesharper/LightInject.Microsoft.DependencyInjection where this behavior is implemented using "child container".

**Comment by geirsagberg** [comment]
Yeah I thought I had it working, but I might have encountered a bug when attempting to use multiple parallel scopes inside a scope, using `PerLogicalCallContextScopeManagerProvider` and async tasks:

```csharp
    public class PerLogicalCallContextTests : IDisposable
    {
        private readonly ServiceContainer container;
        private readonly Scope scope;

        public PerLogicalCallContextTests()
        {
            container = new ServiceContainer();
            container.Register<IFoo, Foo>(new PerScopeLifetime());
            container.ScopeManagerProvider = new PerLogicalCallContextScopeManagerProvider();

            scope = container.BeginScope();
        }

        private void DoInNewScope(Action action)
        {
            using (container.BeginScope()) {
                action();
            }
        }

        [Fact]
        public void Will_get_different_instances_in_different_scopes_async()
        {
            IFoo foo1 = null;
            IFoo foo2 = null;

            Task.WaitAll(
                Task.Run(() => DoInNewScope(() => foo1 = container.GetInstance<IFoo>())),
                Task.Run(() => DoInNewScope(() => foo2 = container.GetInstance<IFoo>()))
            );

            Assert.NotSame(foo1, foo2);
        }

        public void Dispose()
        {
            scope.Dispose();
        }
    }
```

**Comment by seesharper** [comment]
Even though you seemingly start two tasks in parallell here, there is no guarantee that they will actually run in parallel unless there is a good reason to do so. Putting a Task.Delay within the DoInNewScope should make the test pass.
```
        private void DoInNewScope(Action action)
        {
            using (container.BeginScope())
            {
                Task.Delay(100).Wait();
                action();
            }
        }
```

This "ensures" that two threads are started and any subsequent call to GetInstance across await point will return the same instance.

**Comment by geirsagberg** [comment]
I tired your fix, the test still fails. Does it pass for you?

I also tried using `async/await`, to no avail:

```csharp
        private async Task DoInNewScope(Action action)
        {
            using (container.BeginScope()) {
                await Task.Delay(100);
                action();
            }
        }

        [Fact]
        public async Task Will_get_different_instances_in_different_scopes_async()
        {
            IFoo foo1 = null;
            IFoo foo2 = null;

            await Task.WhenAll(
                Task.Run(() => DoInNewScope(() => foo1 = container.GetInstance<IFoo>())),
                Task.Run(() => DoInNewScope(() => foo2 = container.GetInstance<IFoo>()))
            );

            Assert.NotSame(foo1, foo2);
        }
```

**Comment by seesharper** [comment]
Sorry, my bad. You will need to "Wait" for the task since this is not an async method :)

```
        private void DoInNewScope(Action action)
        {
            using (container.BeginScope())
            {
                Task.Delay(100).Wait();
                action();
            }
        }
```

**Comment by geirsagberg** [comment]
Yeah I tried that as well, test still fails. Is it green for you?

**Comment by seesharper** [comment]
That is strange. Just to be sure. Here is the complete test that passes on my side 
```
    public class PerLogicalCallContextTests : IDisposable
    {
        private readonly ServiceContainer container;
        private readonly Scope scope;

        public PerLogicalCallContextTests()
        {
            container = new ServiceContainer();
            container.Register<IFoo, Foo>(new PerScopeLifetime());
            container.ScopeManagerProvider = new PerLogicalCallContextScopeManagerProvider();

            scope = container.BeginScope();
        }

        private void DoInNewScope(Action action)
        {
            using (container.BeginScope())
            {
                Task.Delay(100).Wait();
                action();
            }
        }

        [Fact]
        public void Will_get_different_instances_in_different_scopes_async()
        {
            IFoo foo1 = null;
            IFoo foo2 = null;

            Task.WaitAll(
                Task.Run(() => DoInNewScope(() => foo1 = container.GetInstance<IFoo>())),
                Task.Run(() => DoInNewScope(() => foo2 = container.GetInstance<IFoo>()))
            );

            Assert.NotSame(foo1, foo2);
        }

        public void Dispose()
        {
            scope.Dispose();
        }
    }
```

**Comment by geirsagberg** [comment]
Copied and pasted, fails here :\ might be a timing issue?

**Comment by geirsagberg** [comment]
Hmm, the test passes when using the VS test runner, but not when using ReSharper, nor NCrunch

**Comment by seesharper** [comment]
I will look into this tomorrow

**Comment by seesharper** [comment]
Doing some major brain surgery with scoping. See #314

**Comment by seesharper** [comment]
@geirsagberg Could you give me the complete test using async/await?

**Comment by geirsagberg** [comment]
Here it is, with both tests: https://github.com/geirsagberg/LightInject/blob/CallContextIssue/src/LightInject.Tests/PerLogicalCallContextTests.cs

The ReSharper runner is constantly red, NCrunch is mostly red but sometimes green, and VS Test runner is green...

**Comment by seesharper** [comment]
Could you pull down the latest from master and verify that it works on your side?
Really appreciate the feedback and contributions  👍

**Comment by seesharper** [comment]
Could you pull down the latest from master and verify that it works on your side?
Really appreciate the feedback and contributions  👍

**Comment by geirsagberg** [comment]
All tests green :) Very nice work!
